math/math_helper.go

```diff
@@ -43,6 +50,28 @@ var squareRootTable = map[uint64]uint64{
 	4194304: 2048,
 }
 
+// SquareRootEffectiveBalance implements Newton's algorithm to compute the square root of
+// the given uint64 starting from the last cached value
+func SquareRootEffectiveBalance(balance uint64) uint64 {
+	if balance == 0 {
+		return 0
+	}
+	cachedSquareRoot.Lock()
+	defer cachedSquareRoot.Unlock()
+	if balance == cachedSquareRoot.balance {
+		return cachedSquareRoot.squareRoot
+	}
```

Comment:
If we expect a high throughput of reads vs writes, we could open a RLock here like this:
```suggestion
	cachedSquareRoot.RLock() 
	if balance == cachedSquareRoot.balance {
		cachedSquareRoot.RUnlock()
		return cachedSquareRoot.squareRoot
	}
	cachedSquareRoot.RUnlock()
	cachedSquareRoot.Lock()
	defer cachedSquareRoot.Unlock()
```

Reply:
Nah, I think it's better a simple mutex which is much better performant and this function is not called more than a couple of times per slot, it takes only nanoseconds and it's called in sequence, so never at risk of a race. Only if we sync a block at epoch transition